### ISSUE TITLE
Doesn't apply the dictionary order (sequence) from config file

### ISSUE THREAD

**Issue by GD-fix** [xiaoyifang/goldendict-ng#1128]
**Describe the bug**
The dictionary order (sequence) from existing (previous versions created) config file doesn't apply in new versions (but remains correct in config file at least without doing reorder in new versions).

**To Reproduce**
1. Launch some older version, go to Menu / Edit / Dictionaries and in Dictionaries / Dictionary order do manual reorder of dictionaries in some sequence (see the first screenshot for example).
2. Click on 'OK'.
3. Quit GD-ng.
4. Launch some newer version (for reliability with some dictionaries autoreindexing during startup), go to Menu / Edit / Dictionaries and in Dictionaries / Dictionary order see error (see the second screenshot and Additional context for example).

**Expected behavior**
The dictionaries order (sequence) **must** be applicable from existing previous versions config file. In [used by me dictionaries collection](https://booktracker.org/viewtopic.php?t=73067) on stand alone PC there is > 2 thousand dictionaries and all (hundreds) of the ZIM dictionaries (some of which were in correctly order remained from original GD and some new added in manual order both in system GD-ng version 23.05.03-alpha. on Sun Aug 13 01:30:12 2023) in AppImage release versions were appeared at the end of the dictionaries list in alphabetical order. This prevents me from upgrading to the GD-ng from original GD...

**Screenshots**
![Screenshot_dic_order](https://github.com/xiaoyifang/goldendict-ng/assets/21008593/bc6489e3-eb7c-4dc3-bb5e-8a374cf1686f)
![Screenshot_dic_order_new](https://github.com/xiaoyifang/goldendict-ng/assets/21008593/75d1627d-eba3-4044-b32f-0eb18deec857)

**OS and software versions**
- OS: OpenSUSE Tumbleweed
- Version up to date

**Additional context**
In the config file:
`<dictionaryOrder id="0" name="">
  <dictionary name="Alpine Linux Wiki">4f4fe65ebd60d024131ff36bfb9613fa</dictionary>
  <dictionary name="MultiTran Combinatory × (En-En)">e03ab8a73ed348001b63e164ff25e765</dictionary>
  <dictionary name="ArchWiki">3ba8ed5a0579083b8cb4ecb76cf87e5f</dictionary>
  <dictionary name="Freelang × (Ru-En)">a95e48e8eae30d1dadadc14d851c22f2</dictionary>
  <mutedDictionaries/>
 </dictionaryOrder>`

**Comment by xiaoyifang** [comment]
I guess the dictionary's id has changed in the gd-ng.


fixes ***@***.***> 于2023年9月9日周六 17:09写道：

**Comment by xiaoyifang** [comment]
> 1. Launch some older version,

could the dictionary order be persisted when you use the newer version alone.

**Comment by GD-fix** [comment]
If I don't change the order in newer version, the order in config file doesn't change and so in older version it remains correct and in newer it remains incorrect.

Is it possible to do the next: if in newer version id is changing, to place new id in dictionary order before the old id (and if "Remove invalid index on exit" is enabled, the old id in config file will deleted on quit) or do not allow to change id's of used dictionaries at all?

**Comment by xiaoyifang** [comment]
> If I don't change the order in newer version, the order in config file doesn't change and so in older version it remains correct and in newer it remains incorrect.

When you change the order in newer version , will the order keep unchanged  when the new version open it next time?

**Comment by xiaoyifang** [comment]
Yes ,. I know that.Thanks for that.
I'll check what I can do.

**Comment by GD-fix** [comment]
Thanks in advance.

**Comment by xiaoyifang** [comment]
> P. S. There is "segmentation fault" (how I translate "Ошибка сегментирования (образ памяти сброшен на диск)" to English) on output console after each GoldenDict-ng-v23.09.08-WhiteDew.230907.216505c9 quit (after previous release version quit there isn't any output)...

This can be created in a new issue.

**Comment by xiaoyifang** [comment]
I think the dictionary order issue was introduced in this https://github.com/xiaoyifang/goldendict-ng/commit/c664ac315cb4bffe8bdd4b81713a0a5f74b98847

Double check the implementation ,   it looks good.

the original version's logic has some other problem(I guess).
Put the same dictionary in different folders and add the folders to gd should produce this issue.

**Comment by GD-fix** [comment]
> This can be created in a new issue. and need more detail , any other logs. and what's the dictionary which caused the crash.

It happens only after quit. Why the problem must be in any dictionary? Previous version with the same dictionaries doesn't display any problem...

**Comment by GD-fix** [comment]
In original GD (I didn't test it in GD-ng) if using the same dictionary even in the same path (but via symbolic link) it interpreted how another dictionary (why there is the problem of dictionaries collections distributing without reindexing when user hasn't root privileges and so can't mount directories by the needed paths)...

But in original GD there wasn't any problem with dictionary order remaining on version upgrade even with dictionary reindexing (in case of indexing mechanism change)...

**Comment by xiaoyifang** [comment]
> But in original GD there wasn't any problem with dictionary order remaining on version upgrade even with dictionary reindexing (in case of indexing mechanism change)...

I tend to think that this is a bug in original GD:
The reproduce steps:
1, put the same dictionary ,take D as dictionary,   in two different folders F1,F2.
2, added the F1,F2 folders to gd.
3, there will be two dictionaries appear in the dictionaries.
4,  create two different dictionary groups for each of the above dictionary.   for example ,the dictionary in F1 in group G1, and the dictionary located in F2 in group G2.
5,  quit the gd and then delete the folder F1.
In such a case ,I would expect the G1 group will have no dictionary .

Actual results:
6, open gd ,  the G1 group still have the dictionary .

**Comment by xiaoyifang** [comment]
After thought:
![image](https://github.com/xiaoyifang/goldendict-ng/assets/105986/516c2f03-b70a-460d-a3aa-006d5a6b7300)

from the screenshot ,I think both versions are gd-ng , one is very old indeed.

**Comment by GD-fix** [comment]
I apologize for the delay in responding (I didn't have time for the internet).
1. With groups there was some other behaviour. How I remember, for example, when upgrading dictionaries, if new version dictionary file had another name, but the dictionary name was the same, in Dictionary order it was added to the end, but in Groups (at least after old version dictionary deleting) new version remained in the old version place. Which, by the way, was good, since often new versions of dictionaries have the same name. **But with GD update without dictionaries update the dictionary order remained the same anyway.**
2. It's called GD-ng (after its appearance in mentioned OS with regular update I found it here on Github):
![GoldenDict-ng](https://github.com/xiaoyifang/goldendict-ng/assets/21008593/8ef877c9-5282-48a3-b3a0-461e9798db07)
![GoldeDict-ng_ver](https://github.com/xiaoyifang/goldendict-ng/assets/21008593/3f0b0cff-edc3-4949-bed1-8d4c68e76026)
By the way, what about GD-ng shell translations support in releases?

Returning to a subject of the issue, to enable the transfer of dictionary collection owners from original to "next generation" GD is it possible to release the next behaviour:
1. If dictionary has the same full path (with dictionary files names) and files checksums summary (according to some of the fastest algorithms, and consider them to be the same in case of missing checksums saved in the configuration file when switching from GD versions that did not support them) as is in existing dictionary order configuration, it remains in the order old place nevertheless of reindexing or not (same or another ID).
2. If not, but has the same dictionary name it places in order after existing in config "namesake".
3. Full path must be defined through symbolic links, if any (and in any case, not only dictionary order).

Thanks in advance.

**Comment by xiaoyifang** [comment]
the order is affected by this modification(https://github.com/xiaoyifang/goldendict-ng/commit/c664ac315cb4bffe8bdd4b81713a0a5f74b98847)
![image](https://github.com/xiaoyifang/goldendict-ng/assets/105986/aaeaf4f8-c065-4210-9e3b-7c275229369b)

This has lead to the bug describe in this comment https://github.com/xiaoyifang/goldendict-ng/issues/1128#issuecomment-1712676242

I think you can take this as a  feature which use checksum(unique id) as the only order element.

**Comment by GD-fix** [comment]
1. How I understand, in GD-ng the dictionary unique ID is not equal to the dictionary files checksum. This issue was created just because new GD-ng version doesn't apply the existing in config file order of not changed dictionaries (i.e. with the same old checksums) due to the new ID's created for them. Am I right?
2. Is it possible to leave the dictionary name (+languages) as the second element in ordering for: `2. If not, but has the same dictionary name it places in order after existing in config "namesake".`?

**Comment by GD-fix** [comment]
P.S. Last release: all the same...
The dictionary name must be leaved as the second element in ordering also for using it when GD-ng updating from old versions without supporting the first element - dictionary files checksum. The using of dictionary ID (that may be changed with GD-ng updating) as element for ordering purpose I don't understand at all...

**Comment by xiaoyifang** [comment]
> 1. the dictionary unique ID is not equal to the dictionary files checksum.

https://github.com/xiaoyifang/goldendict-ng/blob/db0ca76ead9511f027fd3562462c678fb9cefcb0/src/dict/dictionary.cc#L506C1-L540C2

this is the dictionary id calculation.  This logic has not changed .






> The using of dictionary ID (that may be changed with GD-ng updating) as element for ordering purpose

the dictionary ID will not change.

The issue here is that the old version's config file has store an incorrect dictionaryId (maybe during updates, the unqiue dictionary id has changed ,but failed to update in the config file.)


About the above issue ,I think you can reproduce with the old goldendict alone.  
configure the old goldendict as portable application and add the problemic dictionaries ,  then compare the config file.  You should be able to notice that the dictionary Id has changed .

**Comment by GD-fix** [comment]
I'm not a developer, but how I understand there is used the Md5 files checksum algorithm (I don't know how it works with multi-files dictionary). But there is mentioned a dictionary path too. And may be there is some differences for assigning ID's in portable and installable versions due to the differences in paths defining (`// For portable version, we use relative paths`)?

OK. Let's consider the specific example.
1. Clear the contents of the index folder.
2. Run the [installed version](https://user-images.githubusercontent.com/21008593/271764568-3f0b0cff-edc3-4949-bed1-8d4c68e76026.png) of GD-ng.
3. Edit the Dictionary order: ![dic_order_1](https://github.com/xiaoyifang/goldendict-ng/assets/21008593/9f5006d6-a32c-4a68-a640-ab8f1ebda634)
4. Quit GD-ng and get in the config file:
![config_order](https://github.com/xiaoyifang/goldendict-ng/assets/21008593/2bd23a97-4a7d-4b88-9e10-d92635e5667d)
 and in the index folder: 
> /.goldendict/index/3ba8ed5a0579083b8cb4ecb76cf87e5f_FTS_x
> /.goldendict/index/4f4fe65ebd60d024131ff36bfb9613fa_FTS_x
> /.goldendict/index/50a7848bfc750cd21b51e276a582dae4_FTS_x
> /.goldendict/index/40396d22c3533545fa1ee41769d8b18b_FTS_x
> /.goldendict/index/e03ab8a73ed348001b63e164ff25e765_FTS_x
> /.goldendict/index/3ba8ed5a0579083b8cb4ecb76cf87e5f
> /.goldendict/index/4f4fe65ebd60d024131ff36bfb9613fa
> /.goldendict/index/50a7848bfc750cd21b51e276a582dae4
> /.goldendict/index/40396d22c3533545fa1ee41769d8b18b
> /.goldendict/index/e03ab8a73ed348001b63e164ff25e765
5. Run GD-ng last portable release.
6. Get the Dictionary order:
![dic_order_2](https://github.com/xiaoyifang/goldendict-ng/assets/21008593/d19ff506-1e1c-4495-bad9-c3a01ed7fcba)
7. Quit GD-ng and get unchanged dictionaryOrder in config file and changed index folder:
> /.goldendict/index/2d25c5fa712521b2e7083de2f81652b1_FTS_x
> /.goldendict/index/3ba8ed5a0579083b8cb4ecb76cf87e5f_FTS_x
> /.goldendict/index/4f4fe65ebd60d024131ff36bfb9613fa_FTS_x
> /.goldendict/index/50a7848bfc750cd21b51e276a582dae4_FTS_x
> /.goldendict/index/59a431b98edbdac937d7731563a9602d_FTS_x
> /.goldendict/index/253b8d79aa21f78775219bafa7ae6cfa_FTS_x
> /.goldendict/index/40396d22c3533545fa1ee41769d8b18b_FTS_x
> /.goldendict/index/e03ab8a73ed348001b63e164ff25e765_FTS_x
> /.goldendict/index/2d25c5fa712521b2e7083de2f81652b1
> /.goldendict/index/3ba8ed5a0579083b8cb4ecb76cf87e5f
> /.goldendict/index/4f4fe65ebd60d024131ff36bfb9613fa
> /.goldendict/index/50a7848bfc750cd21b51e276a582dae4
> /.goldendict/index/59a431b98edbdac937d7731563a9602d
> /.goldendict/index/253b8d79aa21f78775219bafa7ae6cfa
> /.goldendict/index/40396d22c3533545fa1ee41769d8b18b
> /.goldendict/index/e03ab8a73ed348001b63e164ff25e765

To sum up: the 2 Lingvo dictionaries with unchanged ID's remained in the same order (from config file) and the 3 ZIM dictionaries get the new ID's, add (with this ID's) in index folder and in Dictionary order moved in alphabetical order to the end of the list (without any dictionaryOrder changes in config file), i.e. **some dictionary ID will change and there isn't difference the portable or not version for dictionary order with unchanged ID's**...

**Comment by xiaoyifang** [comment]
Thanks for your patient. I got the idea.
I did use a new strategy to generate the zim dictionary's unique id.

**Comment by GD-fix** [comment]
This leads to some other shortcomings in the operation of the GD-ng. For example, during each startup ZIM dictionaries are indexing (instead of loading)...

Is it possible to add the strategy selection in preferences?

Thanks in advance.

**Comment by xiaoyifang** [comment]
> during each startup ZIM dictionaries are indexing (instead of loading)...

There should be some checking.  Indexing should not happen again... (need check)

**Comment by xiaoyifang** [comment]
> Is it possible to add the strategy selection in preferences?

Not in plan ,right now.

**Comment by GD-fix** [comment]
> > during each startup ZIM dictionaries are indexing (instead of loading)...
> 
> There should be some checking. Indexing should not happen again... (need check)

With indexed dictionaries "indexing" is fast...

**Comment by GD-fix** [comment]
> > Is it possible to add the strategy selection in preferences?
> 
> Not in plan ,right now.

OK. Globally: when it will be guaranteed, that any dictionaries already indexed in previous versions will remain in the order prescribed in the configuration file nevertheless of their reindexing ("reIDing") or not in new versions?

**Comment by GD-fix** [comment]
In [GoldenDict-ng-v23.10.01-alpha.231013.9da4dfd8](https://github.com/xiaoyifang/goldendict-ng/releases/tag/v23.10.01-alpha.9da4dfd8) all the same...

Can You explain, what is the problem simply to add old code for ZIM dictionaries ID determination and add in preferences something like "ZIM dictionaries ID determination for backward compatibility" for user selection?

Thanks in advance.

**Comment by xiaoyifang** [comment]
> Can You explain, what is the problem simply to add old code for ZIM dictionaries ID determination and add in preferences something like "ZIM dictionaries ID determination for backward compatibility" for user selection?

I do not want to give users too much options to choose , instead I want to give them the best options to use directly.

Expose an extra option will introduce new cases which will not benefit most of the users and make the daily usage complex.

For the above described cases,  users can rearrange dictionary (only once) and then they will have the correct orders.

the cases only affect zim dictionary so far,  I do not expect the zim dictionary quantity is big.

**Comment by GD-fix** [comment]
In Advanced preferences, for example, there is option "Dictionary debug". You think, that this option will use more users than option, that will free from the need to rearrange a ~**hundred** ("not a big quantity") of dictionaries? 
I had rearranged the dictionaries once in mentioned GoldenDict-ng version from OS main repository. You tell me, that I must rearranged they once again in newer versions. And what guarantees do You give, that this will not be needed in any future version?

Thanks in advance.

**Comment by GD-fix** [comment]
And is the next GD-ng each running output right that You expect:

>Load done
> 
> starting create the fts with thread: 2
> [FULLTEXT] make fts for the dictionary: "Freelang × (Ru-En)"
> [FULLTEXT] make fts for the dictionary: "MultiTran Combinatory × (En-En)"
> .../.goldendict/index/50a7848bfc750cd21b51e276a582dae4_FTS_x dehsinif false
> [FULLTEXT] make fts for the dictionary: "3D Printing"
> .../.goldendict/index/253b8d79aa21f78775219bafa7ae6cfa_FTS_x dehsinif false
> [FULLTEXT] make fts for the dictionary: "Alpine Linux Wiki"
> .../.goldendict/index/59a431b98edbdac937d7731563a9602d_FTS_x dehsinif false
> waiting for all the fts creation to finish.
> [FULLTEXT] make fts for the dictionary: "ArchWiki"
> .../.goldendict/index/2d25c5fa712521b2e7083de2f81652b1_FTS_x dehsinif false
> .../.goldendict/index/e03ab8a73ed348001b63e164ff25e765_FTS_x dehsinif false
> finished/cancel all the fts creation
> reset inspector
> getResource: "gdlookup://localhost?word=Welcome!&group=4294967295&word=&group=4294967294"
> scheme: "gdlookup"
> host: "localhost"
> In-place finish.
> 
> ====reading  13020 of (13020) bytes . Finished: 1
> article view loaded url: "data:text/html;charset=UTF-8,%3C%21DOCTYPE html%3E%0A%3Chtml%3E%3Chead%3E%0A%3Cmeta charset%3D%22utf-8%22%3E%0A%3Cscript src%3D%22qrc%3A%2F%2F%2Fscripts%2Fjquery-3.6.0.slim.min.js%22%3E%3C%2Fscript%3E" true
> article view loaded url: "gdlookup://localhost/?word=Welcome!&group=4294967295" true

?

**Comment by xiaoyifang** [comment]
> [FULLTEXT] make fts for the dictionary: "Freelang × (Ru-En)"
> [FULLTEXT] make fts for the dictionary: "MultiTran Combinatory × (En-En)"
> .../.goldendict/index/50a7848bfc750cd21b51e276a582dae4_FTS_x dehsinif false
> [FULLTEXT] make fts for the dictionary: "3D Printing"
> .../.goldendict/index/253b8d79aa21f78775219bafa7ae6cfa_FTS_x dehsinif false
> [FULLTEXT] make fts for the dictionary: "Alpine Linux Wiki"
> .../.goldendict/index/59a431b98edbdac937d7731563a9602d_FTS_x dehsinif false
> waiting for all the fts creation to finish.
> [FULLTEXT] make fts for the dictionary: "ArchWiki"
> .../.goldendict/index/2d25c5fa712521b2e7083de2f81652b1_FTS_x dehsinif false
> .../.goldendict/index/e03ab8a73ed348001b63e164ff25e765_FTS_x dehsinif false

these are some debug messages only for debug .  they are expected when checking the integrity of dictionary.

**Comment by xiaoyifang** [comment]
> there is option "Dictionary debug"

This is an experimental option and maybe delete in the future.

> that will free from the need to rearrange a ~**hundred** ("not a big quantity") of dictionaries?

I think only the zim dictionaries need to be rearranged. 

> And what guarantees do You give, that this will not be needed in any future version?
> 
> Thanks in advance.

Without good reasons ,the dictionaryid logic should be kept the same and will never be changed .

**Comment by xiaoyifang** [comment]
In case you have fasle expectation .

I'll close this ,  this issue is unlikely to be fixed in in the near future

**Comment by GD-fix** [comment]
> 
> 
>     that will free from the need to rearrange a ~hundred ("not a big quantity") of dictionaries?
> 
> I think only the zim dictionaries need to be rearranged.
> 
>     And what guarantees do You give, that this will not be needed in any future version?
> 
>     Thanks in advance.
> 
> Without good reasons ,the zim dictionary's unique id should be kept the same logic and will never be changed .

Yes, "only" a ~hundred ZIM dictionaries, not a thousands all dictionaries...

So my question is exactly that, what guarantees do You give, that there aren't and will not be any "good reasons" to rearranged the ZIM dictionaries (if already, at least once, it took place to be)?

Thanks in advance.

**Comment by GD-fix** [comment]
> these are some debug messages only for debug . they are expected when checking the integrity of dictionary's fulltext index.

I ask You to clarify, these debug messages took place to be only when GD-ng starting from console, and in usual case during a thousands dictionaries loading it doesn't take extra time and memory?
